### PR TITLE
Hz 0304 add property

### PR DESCRIPTION
**Pull request by Kelozheng**
HZ - 304 add property and get property function

### CODE REVIEW

File: src/main/java/com/eta/houzezbackend/mapper/AgentMapper.java [hunk 1]
```diff
@@ -19,4 +20,7 @@ public interface AgentMapper {
     @Mapping(target = "activated", constant = "false")
     Agent agentSignUpDtoToAgent(AgentSignUpDto agentSignUpDto);
 
+    Agent agentGetDtoToAgent(AgentGetDto agentGetDto);
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/mapper/AgentMapper.java:23]: Do we need this?

File: src/main/java/com/eta/houzezbackend/model/Agent.java [hunk 2]
```diff
@@ -6,9 +6,12 @@
 import org.springframework.data.jpa.domain.support.AuditingEntityListener;
 
 import javax.persistence.*;
+import java.util.ArrayList;
 import java.util.Date;
+import java.util.List;
 
 @Entity
+@Table(name = "agent")
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/model/Agent.java:14]: Do we need this?

File: src/main/java/com/eta/houzezbackend/repository/CorsFilter.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.springframework.context.annotation.Configuration;
+import javax.servlet.*;
+import javax.servlet.annotation.WebFilter;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@WebFilter(filterName = "CorsFilter ")
+@Configuration
+public class CorsFilter implements Filter {
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/repository/CorsFilter.java:11]: Do we need this?

File: src/main/java/com/eta/houzezbackend/util/Type.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.eta.houzezbackend.util;
+
+public enum Type {
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/util/Type.java:3]: PropertyType

File: build.gradle [hunk 5]
```diff
@@ -35,6 +35,8 @@ dependencies {
 
     implementation 'org.mapstruct:mapstruct:1.5.2.Final'
     implementation 'junit:junit:4.13.2'
+    implementation 'javax.xml.bind:jaxb-api:2.3.0'
```
Comment (yuanningd) [inline build.gradle:38]: what is this for

File: src/main/java/com/eta/houzezbackend/dto/PropertyGetDto.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class PropertyGetDto {
+
+    private long id;
+    private Type type;
+
+    private String title;
+    private Integer price;
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/dto/PropertyGetDto.java:19]: why sometimes use boxing type sometimes use unboxed type

File: src/main/java/com/eta/houzezbackend/service/PropertyService.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    private final PropertyRepository propertyRepository;
+    private final PropertyMapper propertyMapper;
+
+    public PropertyGetDto createNewProperty(PropertyCreateDto propertyCreateDto) {
+        Property property = propertyMapper.propertyCreateDtoToProperty(propertyCreateDto);
+        try {
+            propertyRepository.save(property);
+        } catch (DataIntegrityViolationException e) {
+            if (e.getMostSpecificCause().getClass().getName().equals("org.postgresql.util.PSQLException"))
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/service/PropertyService.java:28]: do we need this?

File: src/main/java/com/eta/houzezbackend/dto/AgentSignUpDto.java [hunk 8]
```diff
@@ -16,8 +15,6 @@
 @Builder
 public class AgentSignUpDto {
 
-    @NotBlank(message = "Email must not be blank.")
-    @Email(message = "Email should be a valid email.")
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/dto/AgentSignUpDto.java:20]: why do you delete this?

File: src/main/java/com/eta/houzezbackend/mapper/AgentMapper.java (hunk 1 above)
Comment (Kelozheng) [reply src/main/java/com/eta/houzezbackend/mapper/AgentMapper.java:23]: need for controller test

File: src/main/java/com/eta/houzezbackend/repository/CorsFilter.java (hunk 3 above)
Comment (Kelozheng) [reply src/main/java/com/eta/houzezbackend/repository/CorsFilter.java:11]: I didn't add this

File: src/main/java/com/eta/houzezbackend/service/PropertyService.java (hunk 7 above)
Comment (Kelozheng) [reply src/main/java/com/eta/houzezbackend/service/PropertyService.java:28]: will delete try catch

File: src/main/java/com/eta/houzezbackend/config/SecurityConfig.java [hunk 9]
```diff
@@ -34,6 +34,9 @@ public class SecurityConfig {
     private static final String[] AUTH_URL_WHITELIST = {
             "/agents",
             "/agents/sign-in",
+            "/properties/*",
+            "/agents/*/properties/get-propertylistbyagent",
+            "/agents/*/properties/create-property",
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/config/SecurityConfig.java:39]: ？

File: src/main/java/com/eta/houzezbackend/controller/AgentController.java [hunk 10]
```diff
@@ -55,4 +60,16 @@ public void getAgentByEmail(@RequestParam String email) {
...
         agentService.resendEmail(map.get("email"));
     }
+
+    @PostMapping("/{id}/properties/create-property")
+    @ResponseStatus(HttpStatus.CREATED)
+    public PropertyGetDto addProperty(@Valid @RequestBody PropertyCreateDto propertyCreateDto, @PathVariable long id) {
+        return propertyService.createNewProperty(propertyCreateDto, id);
+    }
+
+    @GetMapping("/{id}/properties/get-propertylistbyagent")
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/controller/AgentController.java:70]: still use verb?

File: src/main/java/com/eta/houzezbackend/exception/ControllerExceptionHandler.java [hunk 11]
```diff
@@ -42,6 +42,12 @@ public ErrorDto handleUniqueEmailViolationException(UniqueEmailViolationExceptio
         return new ErrorDto(HttpStatus.UNPROCESSABLE_ENTITY.getReasonPhrase(), List.of(Objects.requireNonNull(e.getMessage())));
     }
 
+    @ExceptionHandler(value = {UniquePropertyViolationException.class})
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/exception/ControllerExceptionHandler.java:45]: what is this for

File: src/main/java/com/eta/houzezbackend/repository/PropertyRepository.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.eta.houzezbackend.model.Property;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+
+public interface PropertyRepository extends JpaRepository<Property, Long> {
+    @Query(value = "select * from property as a left join agent as b on a.agent_id=b.id", nativeQuery = true)
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/repository/PropertyRepository.java:11]: use jpql instead

**Comment by Kelozheng** [comment]
For test function. Will delete when merge

File: src/main/java/com/eta/houzezbackend/service/PropertyService.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Transactional
+    public PropertyGetDto getProperty(Long id) {
+        Property property = propertyRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Property", id));
+        return propertyMapper.propertyToPropertyGetDto(property);
+    }
+
+
+    public ResponseEntity<Map<String, Object>> getPropertiesByAgent(long id, int page, int size) {
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/service/PropertyService.java:49]: why do we have ResponseEntity here

File: src/main/java/com/eta/houzezbackend/service/PropertyService.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            Pageable paging = PageRequest.of(page, size);
+            Page<Property> pageTuts = propertyRepository.findAByAgentOrderById(id, paging);
+            List<PropertyGetDto> properties = pageTuts.getContent().stream().map(propertyMapper::propertyToPropertyGetDto).collect(Collectors.toList());
+
+            if (properties.isEmpty()) {
+                return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+            }
+
+            Map<String, Object> response = new HashMap<>();
```
Comment (yuanningd) [inline src/main/java/com/eta/houzezbackend/service/PropertyService.java:60]: why use hash map instead of dto?

**Comment by Kelozheng** [comment]
Want to send the total page number and the list to front end at the same
time

**Comment by Kelozheng** [comment]
Can I use /new-propety for add
/propertylist for get?

**Review by yuanningd — APPROVED** [review]

**Pull request merged**